// mcporter@0.10.2/dist/runtime/oauth.js
import { isUnauthorizedError } from '../runtime-oauth-support.js';
export const DEFAULT_OAUTH_CODE_TIMEOUT_MS = 300_000;
const OAUTH_FLOW_ERROR = Symbol('oauth-flow-error');
const POST_AUTH_CONNECT_ERROR = Symbol('post-auth-connect-error');
const MAX_OAUTH_ERROR_DETAIL_LENGTH = 1_200;
export class OAuthTimeoutError extends Error {
    timeoutMs;
    serverName;
    constructor(serverName, timeoutMs) {
        const seconds = Math.round(timeoutMs / 1000);
        super(`OAuth authorization for '${serverName}' timed out after ${seconds}s; aborting.`);
        this.name = 'OAuthTimeoutError';
        this.timeoutMs = timeoutMs;
        this.serverName = serverName;
    }
}
export class OAuthAuthorizationNotStartedError extends Error {
    serverName;
    constructor(serverName, cause) {
        const causeMessage = formatOAuthErrorDetail(cause);
        const detail = causeMessage ? ` Last error: ${causeMessage}` : '';
        super(`OAuth authorization for '${serverName}' did not produce an authorization URL; aborting instead of waiting for a browser callback.${detail}`);
        this.name = 'OAuthAuthorizationNotStartedError';
        this.serverName = serverName;
    }
}
function formatOAuthErrorDetail(cause) {
    if (!(cause instanceof Error) || !cause.message) {
        return '';
    }
    return truncateOAuthErrorDetail(cause.message);
}
function truncateOAuthErrorDetail(message) {
    if (message.length <= MAX_OAUTH_ERROR_DETAIL_LENGTH) {
        return message;
    }
    const truncated = message.length - MAX_OAUTH_ERROR_DETAIL_LENGTH;
    return `${message.slice(0, MAX_OAUTH_ERROR_DETAIL_LENGTH)}... [truncated ${truncated} chars]`;
}
export function markOAuthFlowError(error) {
    return markError(error, OAUTH_FLOW_ERROR);
}
export function isOAuthFlowError(error) {
    return hasErrorMarker(error, OAUTH_FLOW_ERROR);
}
export function markPostAuthConnectError(error) {
    return markError(error, POST_AUTH_CONNECT_ERROR);
}
export function isPostAuthConnectError(error) {
    return hasErrorMarker(error, POST_AUTH_CONNECT_ERROR);
}
function markError(error, marker) {
    if (!error || (typeof error !== 'object' && typeof error !== 'function')) {
        return error;
    }
    Object.defineProperty(error, marker, {
        value: true,
        enumerable: false,
        configurable: true,
    });
    return error;
}
function hasErrorMarker(error, marker) {
    return (!!error &&
        (typeof error === 'object' || typeof error === 'function') &&
        marker in error &&
        Boolean(error[marker]));
}
export async function connectWithAuth(client, transport, session, logger, options = {}) {
    const { serverName, maxAttempts = 3, oauthTimeoutMs = DEFAULT_OAUTH_CODE_TIMEOUT_MS, recreateTransport } = options;
    const state = {
        activeTransport: transport,
        attempt: 0,
        hasCompletedAuthFlow: false,
    };
    while (true) {
        try {
            return await attemptTransportConnect(client, state);
        }
        catch (error) {
            const unauthorized = isUnauthorizedError(error);
            if (!shouldRetryAuthorization(state, unauthorized, session)) {
                await closeReplacementTransport(transport, state.activeTransport);
                throw state.hasCompletedAuthFlow && !unauthorized ? markPostAuthConnectError(error) : error;
            }
            state.attempt += 1;
            if (state.attempt > maxAttempts) {
                await closeReplacementTransport(transport, state.activeTransport);
                throw state.hasCompletedAuthFlow ? markPostAuthConnectError(error) : error;
            }
            if (session.hasAuthorizationRedirectStarted?.() !== false) {
                logger.warn(`OAuth authorization required for '${serverName ?? 'unknown'}'. Waiting for browser approval...`);
            }
            try {
                state.activeTransport = await completeAuthorizationChallenge(state.activeTransport, session, logger, error, {
                    serverName,
                    oauthTimeoutMs,
                    recreateTransport,
                });
                state.hasCompletedAuthFlow = true;
                logger.info('Authorization code accepted. Retrying connection...');
            }
            catch (authError) {
                const message = authError instanceof OAuthAuthorizationNotStartedError
                    ? 'OAuth authorization could not start.'
                    : 'OAuth authorization failed while waiting for callback.';
                logger.error(message, authError);
                await closeReplacementTransport(transport, state.activeTransport);
                throw markOAuthFlowError(authError);
            }
        }
    }
}
async function attemptTransportConnect(client, state) {
    await client.connect(state.activeTransport);
    return state.activeTransport;
}
function shouldRetryAuthorization(_state, unauthorized, session) {
    if (!session || !unauthorized) {
        return false;
    }
    return true;
}
async function closeReplacementTransport(originalTransport, activeTransport) {
    if (activeTransport === originalTransport) {
        return;
    }
    await activeTransport.close().catch(() => { });
}
async function completeAuthorizationChallenge(transport, session, logger, connectError, options) {
    if (session.hasAuthorizationRedirectStarted?.() === false) {
        throw new OAuthAuthorizationNotStartedError(options.serverName ?? 'unknown', connectError);
    }
    const code = await waitForAuthorizationCodeWithTimeout(session, logger, options.serverName, options.oauthTimeoutMs ?? DEFAULT_OAUTH_CODE_TIMEOUT_MS);
    if (typeof transport.finishAuth !== 'function') {
        logger.warn('Transport does not support finishAuth; cannot complete OAuth flow automatically.');
        throw connectError;
    }
    await transport.finishAuth(code);
    if (!options.recreateTransport) {
        return transport;
    }
    const nextTransport = await options.recreateTransport(transport);
    await transport.close().catch(() => { });
    return nextTransport;
}
// Race the pending OAuth browser handshake so the runtime can't sit on an unresolved promise forever.
export function waitForAuthorizationCodeWithTimeout(session, logger, serverName, timeoutMs = DEFAULT_OAUTH_CODE_TIMEOUT_MS) {
    if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
        return session.waitForAuthorizationCode();
    }
    const displayName = serverName ?? 'unknown';
    return new Promise((resolve, reject) => {
        const timer = setTimeout(() => {
            const error = new OAuthTimeoutError(displayName, timeoutMs);
            logger.warn(error.message);
            reject(error);
        }, timeoutMs);
        session.waitForAuthorizationCode().then((code) => {
            clearTimeout(timer);
            resolve(code);
        }, (error) => {
            clearTimeout(timer);
            reject(error);
        });
    });
}
export function parseOAuthTimeout(raw) {
    if (!raw) {
        return DEFAULT_OAUTH_CODE_TIMEOUT_MS;
    }
    const parsed = Number.parseInt(raw, 10);
    if (!Number.isFinite(parsed) || parsed <= 0) {
        return DEFAULT_OAUTH_CODE_TIMEOUT_MS;
    }
    return parsed;
}
export function resolveOAuthTimeoutFromEnv() {
    return parseOAuthTimeout(process.env.MCPORTER_OAUTH_TIMEOUT_MS ?? process.env.MCPORTER_OAUTH_TIMEOUT);
}
