// mcporter@0.10.2/dist/daemon/launch.js
import { spawn } from 'node:child_process';
import path from 'node:path';
export function launchDaemonDetached(options) {
    const invocation = buildDaemonLaunchInvocation(options);
    const child = spawn(invocation.command, invocation.args, {
        detached: true,
        stdio: 'ignore',
        env: invocation.env,
    });
    child.unref();
}
export function buildDaemonLaunchInvocation(options, processInfo = {
    argvEntry: process.argv[1],
    env: process.env,
    execArgv: process.execArgv,
    execPath: process.execPath,
    platform: process.platform,
}) {
    const cliEntry = resolveCliEntry(processInfo.argvEntry);
    const configArgs = options.configExplicit ? ['--config', options.configPath] : [];
    const args = [
        ...processInfo.execArgv,
        ...(cliEntry ? [cliEntry] : []),
        ...configArgs,
        ...(options.rootDir ? ['--root', options.rootDir] : []),
        'daemon',
        'start',
        '--foreground',
        ...(options.extraArgs ?? []),
    ];
    const env = {
        ...processInfo.env,
        MCPORTER_DAEMON_CHILD: '1',
        MCPORTER_DAEMON_SOCKET: options.socketPath,
        MCPORTER_DAEMON_METADATA: options.metadataPath,
    };
    if (shouldWrapDetachedLaunchWithNohup(processInfo.platform, cliEntry)) {
        return {
            command: 'nohup',
            args: [processInfo.execPath, ...args],
            env,
        };
    }
    return {
        command: processInfo.execPath,
        args,
        env,
    };
}
function shouldWrapDetachedLaunchWithNohup(platform, cliEntry) {
    return platform === 'darwin' && cliEntry === undefined;
}
function resolveCliEntry(entry = process.argv[1]) {
    if (!entry) {
        throw new Error('Unable to resolve mcporter entry script.');
    }
    // In Bun compiled binaries, argv[1] is a virtual /$bunfs/... path that Bun
    // auto-injects into every spawned child.  Including it explicitly would
    // duplicate it and break CLI argument parsing in the child process.
    if (entry.startsWith('/$bunfs/')) {
        return undefined;
    }
    return path.resolve(entry);
}
